pkg/cloudproviders/k8s/k8s_test.go

```diff
@@ -0,0 +1,283 @@
+
+	options := gostatsd.CloudProviderOptions{
+		Viper:    v,
+		Logger:   logrus.StandardLogger(),
+		Version:  "test",
+		NodeName: nn,
+	}
+	cloudProvider, err := NewProvider(options, fakeClient)
+	if err != nil {
+		return err
+	}
+
+	r, ok := cloudProvider.(gostatsd.Runner)
+	assert.True(t, ok)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		r.Run(ctx) // fork
+	}()
+
+	test(t, &testFixtures{
+		fakeClient:    fakeClient,
+		cloudProvider: cloudProvider,
+		podsWatch:     podsWatch,
+	})
+	return nil
+}
+
+type tagTest struct {
+	name                 string
+	viperParams          map[string]interface{}
+	pods                 []*core_v1.Pod
+	expectedNumTags      int
+	expectedTagsContains []tagTestValue
+}
+
+var ipTagTests = []tagTest{
+	{
+		name:                 "TestTagsForIPWithDefaultTagWhitelists",
+		pods:                 []*core_v1.Pod{pod()},
+		expectedNumTags:      1,
+		expectedTagsContains: []tagTestValue{annotationTestValues[0]},
+	},
+	// The same test as above but with two pods to choose from
+	{
+		name:                 "TestTagsForIPWhenMultiplePods",
+		pods:                 []*core_v1.Pod{pod(), pod2()},
+		expectedNumTags:      1,
+		expectedTagsContains: []tagTestValue{annotationTestValues[0]},
+	},
+	{
+		name: "TestTagsForIPWithNoAnnotationTagWhitelist",
+		viperParams: map[string]interface{}{
+			FullyQualifiedAnnotationWhitelistParam: []string{},
+		},
+		pods:            []*core_v1.Pod{pod()},
+		expectedNumTags: 0,
+	},
+	{
+		name: "TestTagsForIPWithCustomAnnotationTagWhitelist",
+		viperParams: map[string]interface{}{
+			FullyQualifiedAnnotationWhitelistParam: []string{"^product.company.com/"},
+		},
+		pods:                 []*core_v1.Pod{pod()},
+		expectedNumTags:      1,
+		expectedTagsContains: []tagTestValue{annotationTestValues[1]},
+	},
+	{
+		name: "TestTagsForIPWithCustomLabelTagWhitelist",
+		viperParams: map[string]interface{}{
+			FullyQualifiedLabelWhitelistParam: []string{"^app"},
+		},
+		pods:                 []*core_v1.Pod{pod()},
+		expectedNumTags:      2, // we're still using the default annotation whitelist too
+		expectedTagsContains: []tagTestValue{labelTestValues[0], annotationTestValues[0]},
+	},
+}
+
+func TestIPToTags(t *testing.T) {
+	t.Parallel()
+
+	for _, testCase := range ipTagTests {
+		flags := viper.New()
+		for k, v := range testCase.viperParams {
+			flags.Set(k, v)
+		}
+
+		// Run tests in their own namespace for clarity
+		t.Run(testCase.name, func(tt *testing.T) {
+			err := setupTest(tt, func(ttt *testing.T, fixtures *testFixtures) {
+				for _, p := range testCase.pods {
+					fixtures.podsWatch.Add(p)
+				}
+
+				ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+				defer cancel()
+
+				instanceData, err := fixtures.cloudProvider.Instance(ctx, ipAddr)
+				require.NoError(ttt, err)
+
+				instance := instanceData[ipAddr]
+				assert.NotNil(ttt, instance)
```

Comment:
I generally favor `require` always, however in this case you're depending on `instance` being not-nil (otherwise you get the panic), so this should absolutely be `required`.

Reply:
I learned that the hard way before I got around to reading this 😔.
✅ Always `require` on things that will break further down.